### PR TITLE
[Execution] Output event system

### LINKED ISSUE #2005: Output event system
## Rationale

Allow smart contracts to emit arbitrary messages during execution, that can be seen by outside observers.

A bit like this: http://www2.phys.canterbury.ac.nz/dept/docs/manuals/unix/DEC_5.0a_Docs/HTML/ARH9VBTE/VMCHPXXX.HTM

## output_event syscall to output an event from within a smart contract

```
output_event("json string");
```

* fails if the string is longer than a config-defined max
* costs constant gas
* does not return anything
* is also allowed in readonly execution contexts

When this syscall is called, the following element is added to an `call_output_events: Vec<SCOutputEvent>` that lists all the events generated by the execution of the smart contract, in the order in which they were emitted:

```rust
struct SCOutputEvent {
    "context": Execution context when the call happened: slot, address, block, call stack etc...,
    "data": the emitted JSON string
}
```

## API access

The API must allow listing all the events that happened on the execution of a given ExecuteSC operation in a block, or during a given slot range (up to a config-defined maximum history length), with an optional filter according to the emitter smart contract address.

When running a readonly call, output events are listed as well and sent among the API output

### PR DESCRIPTION
**Pull request by AureliaDolo**
Resolve #2005

### CODE REVIEW

**Comment by AureliaDolo** [comment]
If anyone has better name ideas

**Comment by AureliaDolo** [comment]
Add generate event to sc runtime interface

**Comment by AureliaDolo** [comment]
I did mostly the api part for now. I was thinking of merging this and finishing the work on execution and interface part once #2025 is merged into #1979

**Review by adrien-zinger — APPROVED** [review]
[this issue](https://github.com/massalabs/massa/issues/2035) in massa and [this issue](https://github.com/massalabs/sc-runtime/issues/40) in the runtime are the next steps.

We can merge this part in the feature branch and continue forward the development with [the new interface](https://github.com/massalabs/massa/pull/2025).

File: massa-models/src/output_event.rs
Comment (gterzian) [inline massa-models/src/output_event.rs:19]: I think this has already been done [here](https://github.com/massalabs/massa/blob/399fb6e02a977903d3e085084978c671b19403ba/massa-execution/src/types.rs#L31)? cc @damip

Comment (damip) [reply massa-models/src/output_event.rs:19]: yes indeed

**Comment by AureliaDolo** [comment]
@yvan-sraka please don't merge anything in my branch

**Comment by yvan-sraka** [comment]
Right! Do not hesitate to dismiss my commit :)

File: massa-execution/src/interface_impl.rs
Comment (damip) [inline massa-execution/src/interface_impl.rs:163]: let's not use context_guard! at every line. This causes CPU-wide sync every time and is very heavy. Better lock once at the beginning of the function (see how it was implemented in other methods of InterfaceImpl)

File: massa-execution/src/interface_impl.rs
Comment (damip) [inline massa-execution/src/interface_impl.rs:164]: for now the stack should always be non-empty, but to be future-proof and robust, we did a clean fail in the other methods of InterfaceImpl. Let's do the same here

File: massa-execution/src/interface_impl.rs
Comment (damip) [inline massa-execution/src/interface_impl.rs:170]: Let's remove the entries "bytes_address" and "caller_address" from EventExecutionContext. It's redundant with call_stack

**Review by damip — CHANGES_REQUESTED** [review]

**Review by yvan-sraka — APPROVED** [review]
LGTM

**Review by damip — APPROVED** [review]

**Comment by AureliaDolo** [comment]
bords merge

**Comment by AureliaDolo** [comment]
bors merge

**Comment by damip** [comment]
bors merge

**Comment by damip** [comment]
bors merge

**Comment by damip** [comment]
bors merge

**Pull request merged**
